### PR TITLE
[FLINK-31842][runtime] Calculate the utilization of the task executor only when it is using

### PR DESCRIPTION
**Pull request by huwh**
## What is the purpose of the change
Currently DefaultAllocatedSlotPool#getFreeSlotsInformation always calculates the taskExecutorUtilization. This causes task schedules to be too slow when there are lots of slots, such as 20000 slots total. But only the EvenlySpreadOutLocationPreferenceSlotSelectionStrategy uses this utilization.

## Brief change log

  - *Maintain free slots per task executor in DefaultAllocatedSlotPool to reduce the cost of task executor utilization calculation.*
  - *Calculate task executor's utilization only when used*

## Verifying this change
This change is already covered by existing tests.

#### origin version
```
Benchmark                                                        (jobConfiguration)  Mode  Cnt      Score       Error  Units
RestartStreamingJobBenchmarkExecutor.startSchedulingAndFailover               BATCH    ss    6   7963.471 ?  3040.474  ms/op
RestartStreamingJobBenchmarkExecutor.startSchedulingAndFailover           STREAMING    ss    6  41735.536 ? 13175.253  ms/op
RestartStreamingJobBenchmarkExecutor.startSchedulingAndFailover        BATCH_EVENLY    ss    6   9461.976 ?  3823.653  ms/op
RestartStreamingJobBenchmarkExecutor.startSchedulingAndFailover    STREAMING_EVENLY    ss    6  41957.036 ? 10394.205  ms/op
```
#### Maintain free slots per task executor in DefaultAllocatedSlotPool  
```
Benchmark                                                        (jobConfiguration)  Mode  Cnt      Score      Error  Units
RestartStreamingJobBenchmarkExecutor.startSchedulingAndFailover               BATCH    ss    6   7039.183 ? 2073.795  ms/op
RestartStreamingJobBenchmarkExecutor.startSchedulingAndFailover           STREAMING    ss    6  25766.287 ? 7546.104  ms/op
RestartStreamingJobBenchmarkExecutor.startSchedulingAndFailover        BATCH_EVENLY    ss    6   7701.995 ? 1154.260  ms/op
RestartStreamingJobBenchmarkExecutor.startSchedulingAndFailover    STREAMING_EVENLY    ss    6  29273.547 ? 7111.815  ms/op

```

#### Calculate task executor's utilization only when used
```
Benchmark                                                        (jobConfiguration)  Mode  Cnt      Score      Error  Units
RestartStreamingJobBenchmarkExecutor.startSchedulingAndFailover               BATCH    ss    6   5157.789 ?  674.883  ms/op
RestartStreamingJobBenchmarkExecutor.startSchedulingAndFailover           STREAMING    ss    6  14386.616 ? 3058.114  ms/op
RestartStreamingJobBenchmarkExecutor.startSchedulingAndFailover        BATCH_EVENLY    ss    6   8494.799 ? 1747.083  ms/op
RestartStreamingJobBenchmarkExecutor.startSchedulingAndFailover    STREAMING_EVENLY    ss    6  33175.847 ? 5706.390  ms/op

```

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)

### CODE REVIEW

File: flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/DefaultAllocatedSlotPool.java
Comment (reswqa) [inline flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/DefaultAllocatedSlotPool.java:226]: Can we initialize these two fields here and remove the constructor.

File: flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/SlotInfoWithUtilization.java
Comment (reswqa) [inline flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/SlotInfoWithUtilization.java:86]: I'd prefer remove this method as its only used in test.

File: flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/SlotInfoWithUtilization.java
Comment (reswqa) [inline flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/SlotInfoWithUtilization.java:89]:
```suggestion
            SlotInfo slotInfo, Function<ResourceID, Double> taskExecutorUtilizationLookup) {
```

File: flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/SlotInfoWithUtilization.java
Comment (reswqa) [inline flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/SlotInfoWithUtilization.java:52]: This method should be removed. The rule of thumb is: If we can think of a way to do verification, it is best not to introduce methods that are purely for testing.

File: flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/LocationPreferenceSlotSelectionStrategy.java
Comment (reswqa) [inline flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/LocationPreferenceSlotSelectionStrategy.java:124]: How about change this method to
```
protected abstract double calculateCandidateScore(
            int localWeigh, int hostLocalWeigh, Supplier<Double> taskExecutorUtilizationSupplier);
```
As it stands, passing `SlotInfoAndResources` seems redundant.

**Review by reswqa — COMMENTED** [review]
Thanks @huwh for bring this, the benchmark result show that this is indeed a big optimization! A small question: Does this optimization only have a more obvious effect in the failover scenario?
This change overall looks good to me, I only left some comments, please take a look.

**Comment by reswqa** [comment]
BTW, It seems that the title is incomplete.

**Comment by huwh** [comment]
Thanks @reswqa, I will update this pr asap.

> Does this optimization only have a more obvious effect in the failover scenario?
yes, the slot pool is empty when job first scheduled. So the heavy operation on slot selecting will be skipped.

**Comment by huwh** [comment]
@reswqa I have updated this PR as suggested. PTAL

Migrate some slot selection strategy unit tests to junit5 https://github.com/apache/flink/pull/22424/commits/698428de7816cfd3185562095037179a96f4b119

Address comments: https://github.com/apache/flink/pull/22424/commits/4061dee5b1892265888aa3865bd93673ec16730a, https://github.com/apache/flink/pull/22424/commits/6554923c794455c75723c23f83848452c88731c2

File: flink-runtime/src/test/java/org/apache/flink/runtime/clusterframework/types/LocationPreferenceSlotSelectionStrategyTest.java
Comment (reswqa) [inline flink-runtime/src/test/java/org/apache/flink/runtime/clusterframework/types/LocationPreferenceSlotSelectionStrategyTest.java:168]: The same logic appears in these two test class many times, I'd prefer extracts a specific method to handle this assertion.

File: flink-runtime/src/test/java/org/apache/flink/runtime/clusterframework/types/LocationPreferenceSlotSelectionStrategyTest.java
Comment (reswqa) [inline flink-runtime/src/test/java/org/apache/flink/runtime/clusterframework/types/LocationPreferenceSlotSelectionStrategyTest.java:103]: I'd prefer extract this logic to a specific method, this will also be helpful for adding new test case in the future.

File: flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/SlotInfoWithUtilization.java
Comment (reswqa) [inline flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/SlotInfoWithUtilization.java:48]: It seems that this method has still not been removed. Have you just forgotten it or do you have other concerns 🤔

**Review by reswqa — COMMENTED** [review]
Thanks @huwh for the quick update! This change overall look good to me, I only left some minor comments, please take a look in your free time.

**Comment by huwh** [comment]
@reswqa Thanks for the review. 

Sorry for missing one comments. Now it's updated. 


extract some logic in SlotSelectionStrategy tests: https://github.com/apache/flink/pull/22424/commits/1fed03786d24d11f031489a8ba8c59a5a0459600

migrate some slot pool related tests to junit5, this will be squash to the first hotfix commit and change the commit message before merge: https://github.com/apache/flink/pull/22424/commits/b7d852a3621a1eefe6efe6219303f0ef4854a8ea

remove function getTaskExecutorUtilization which is only used in test: https://github.com/apache/flink/pull/22424/commits/5807589fa15e92c399ea6585d9663e3fe0e7bbb1

**Review by reswqa — APPROVED** [review]
Thanks @huwh for the quick update, LGTM.
To be on the safe side, let's squash all fix-up commits and wait for @wanglijie95 to do the final round review.

File: flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/DefaultAllocatedSlotPool.java
Comment (wanglijie95) [inline flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/DefaultAllocatedSlotPool.java:226]: It seems that we only need the size of `Set<AllocationID>` here, how about changing it to `Map<ResourceID, Long>` ?

File: flink-runtime/src/test/java/org/apache/flink/runtime/clusterframework/types/SlotSelectionStrategyTestBase.java
Comment (wanglijie95) [inline flink-runtime/src/test/java/org/apache/flink/runtime/clusterframework/types/SlotSelectionStrategyTestBase.java:81]: Is the change(remove the constructor) here is necessary?

File: flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/SlotSelectionStrategy.java
Comment (wanglijie95) [inline flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/SlotSelectionStrategy.java:62]: I think we can directly remove the `SlotInfoAndResources` and replace it with `SlotInfoWithUtilization`. This two class has the same functions.  (this change is better to be a separate commit.)

**Review by wanglijie95 — COMMENTED** [review]
@huwh  Thanks for creating this PR. I left some comments, please take a look.

File: flink-runtime/src/test/java/org/apache/flink/runtime/clusterframework/types/SlotSelectionStrategyTestBase.java
Comment (huwh) [reply flink-runtime/src/test/java/org/apache/flink/runtime/clusterframework/types/SlotSelectionStrategyTestBase.java:81]: Previously, we used different constructors to create different SlotSelectionStrategy. On Junit5, the test class is required to have only one constructor, which causes LocationPreferenceSlotSelectionStrategyTest not working properly. It has two constructors, one to create DefaultLocationPreferenceSlotSelectionStrategy, and one for subclass PreviousAllocationSlotSelectionStrategyTest to create PreviousAllocationSlotSelectionStrategy

So here we change to create different SlotSelectionStrategy by BeforeEach function. The constructor is not used.

**Comment by huwh** [comment]
@wanglijie95  Thanks for review. comments addressed and replied

remove SlotInfoAndResources in https://github.com/apache/flink/pull/22424/commits/887f53a67bff76fbffb10b710efd15cb71db068a

Only maintain free slots number in https://github.com/apache/flink/commit/fc38ab92c9615a3392805c565ddd844d469796e0

File: flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/DefaultAllocatedSlotPool.java
Comment (wanglijie95) [reply flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/DefaultAllocatedSlotPool.java:228]: Although the current implementation may not pass in an already freeed slot, I suggest keeping the same logic as before. A suggestion:
```
if (freeSlotsSince.put(slot.getAllocationId(), currentTime) == null) {
    freeSlotsNumberPerTaskExecutor.merge(slot.getTaskManagerId(), 1, Integer::sum);
}
```

File: flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/DefaultAllocatedSlotPool.java
Comment (wanglijie95) [inline flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/DefaultAllocatedSlotPool.java:239]: I suggest to take advantage of the return value of the `remove` method , and use `compute` mothod here, a suggestion:
```
Long freeSince = freeSlotsSince.remove(slot.getAllocationId());
if (freeSince != null) {
    freeSlotsNumberPerTaskExecutor.computeIfPresent(
            slot.getTaskManagerId(),
            (ignore, count) -> {
                int newCount = count - 1;
                return newCount == 0 ? null : newCount;
            });
}

return freeSince;
```
Please correct me if I'm wrong :)

**Comment by huwh** [comment]
Thanks @wanglijie95 , This is a good suggestion, I have squash the previous fix-up commits and update this pr according to the suggestions.

**Review by wanglijie95 — APPROVED** [review]
LGTM

**Comment by reswqa** [comment]
Thanks @huwh and @wanglijie95, Merged % CI.

**Pull request closed without merging**
